proposals/webidl-bindings/Explainer.md

```diff
@@ -0,0 +1,649 @@
+  create [ES Classes] enabling, e.g., Web Components to be fully implemented in
+  WebAssembly.
+
+Given this overview, the proposal is now presented in a bottom-up fashion.
+
+
+## Web IDL Types
+
+Symmetric to the core WebAssembly [Type Section], all compound Web IDL types are
+defined in a **Web IDL Type Subsection** that is stored inside of, and can be
+referenced throughout the rest of, the new **Web IDL Bindings Section** being
+defined by this proposal. The section/subsection design of Web IDL Bindings
+would mirror that of the [Name Section] and allow for extension in the future.
+
+A tentative list of Web IDL compound types includes:
+* Function: a signature including one of {static, method or constructor},
+  parameter types and return type (used for Web IDL [Operation] and [Callback]
+  types)
+* [Dictionary]: listing field names and field types
+* [Enumeration]: listing string values
+* [Union]: listing the member types
+
+Many compound Web IDL types are missing from the above list because these are
+the only compound types with explicit operators (introduced below) for
+constructing and destructuring values.  New compound types can be added in the
+future in conjunction with new operators.  Until then, types not listed above
+can be passed using either the Web IDL [`any`] type, which implies a runtime type
+check, or, in the future, with a [Type Import][Type Imports] that is checked for
+signature-compatibility at instantiation-time, eliminating any runtime checks.
+(It is an [open question](#open-questions) whether this proposal should depend
+on Type Imports and remove the implicit `any` downcast; the rest of the
+proposal assumes "no".)
+
+Compound type definitions also do not include the [Web IDL name]. The
+reason for this is that the module-declared Web IDL types are matched against
+the actual imported Web IDL functions *structurally*.  For simplicity,
+recursive types are disallowed, allowing simple instantiation-time
+type-compatibility checks.
+
+Also notably absent is any sort of [Interface] definition listing the members
+of an interface.  Due to the lack of polymorphic/virtual dispatch in Web IDL,
+these are unnecessary: WebAssembly modules import Web IDL methods directly as
+concrete functions and call these functions by passing the `this`-object as an
+explicit parameter (as shown in the examples below).
+
+A **Web IDL Type** is either a compound type (in the Web IDL Types Subsection)
+or a Web IDL scalar types like `any` or `long`.  This is symmetric to the
+proposed WebAssembly [`ref`] type. In the future, [Type Imports] could also be
+used as Web IDL Types.
+
+As an example, the Web IDL Dictionary:
+
+```WebIDL
+dictionary Contact {
+  DOMString name;
+  long age;
+};
+```
+
+could be defined in a WebAssembly module with this (strawman text-format)
+statement:
+
+```wasm
+(webidl-type $Contact (dict (field "name" DOMString) (field "age" long)))
```

Comment:
IMO, the text format of this should be as close as possible to the 'official' syntax for webIDL itself.

Reply:
I think consistency with the existing text format is also a useful property, but probably we can discuss this in a follow-up issue and iterate on the explainer.

Reply:
Can we try to uniformly render custom sections in the text format by means of the proposed [custom annotation syntax](https://github.com/WebAssembly/annotations/blob/master/proposals/annotations/Overview.md)? That would prevent all sorts of compatibility and tooling issues.

Reply:
Ah, I forgot all about that.  So, iiuc, the request is to make sure that any new syntax is contained by some `(@webidl ...)`, and then otherwise the `...` can be anything (as long as it follows the basic paren-matching rules)?

Reply:
Within limits. It has to be lexible as Wasm tokens. That covers a wide range of lexical syntax (because Wasm identifiers can be almost anything), but we'd need to extend the proposal slightly if we wanted to allow other kinds of brackets as well or "foreign" punctuation like commas and semicolons. Those wouldn't be a problem. (Though it would be a problem to allow completely different lexing rules, mainly because of parens in comments and strings.)

Reply:
(That said, I wouldn't recommend going wild with annotation syntax but rather keep the spirit of the surrounding Wasm syntax.)

Reply:
Heh, that makes sense, but for the two practical options we're considering here (wat-style or Web IDL), it sounds like we're fine, so the only real requirement is, whatever it is, it is seated in a `(@webidl ...)`, not a `(webidl...)`, right?

Reply:
Yep!

Reply:
Righto, will fix, thanks for pointing that out.